File: sample/src/main/java/com/massivedisaster/adal/sample/feature/location/FragmentLocation.java
/*
 * ADAL - A set of Android libraries to help speed up Android development.
 *
 * Copyright (c) 2017 ADAL
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.massivedisaster.adal.sample.feature.location;

import android.content.Intent;
import android.location.Location;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.massivedisaster.adal.fragment.BaseFragment;
import com.massivedisaster.adal.sample.R;
import com.massivedisaster.location.LocationManager;
import com.massivedisaster.location.listener.OnLocationManager;
import com.massivedisaster.location.utils.LocationError;

public class FragmentLocation extends BaseFragment {

    private LocationManager mLocationManager;

    private Button mBtnGetLocation, mBtnGetLocationUpdates;
    private TextView mTxtInfo, mTxtInfoUpdates;
    private String mLocationUpdates = "";

    @Override
    protected void getFromBundle(Bundle bundle) {
        // Intended.
    }

    @Override
    protected void restoreInstanceState(@Nullable Bundle savedInstanceState) {
        // Intended.
    }

    @Override
    protected int layoutToInflate() {
        return R.layout.fragment_location;
    }

    @Override
    protected void doOnCreated() {
        getActivity().setTitle(R.string.sample_location);

        mBtnGetLocation = findViewById(R.id.btnGetLocation);
        mTxtInfo = findViewById(R.id.txtInfo);

        mBtnGetLocationUpdates = findViewById(R.id.btnGetLocationUpdates);
        mTxtInfoUpdates = findViewById(R.id.txtInfoUpdates);

        initialize();
    }

    public void initialize() {
        mBtnGetLocation.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mTxtInfo.setText("Getting location...");
                v.setEnabled(false);
                getLocation();
            }
        });
        mBtnGetLocationUpdates.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (v.isSelected()) {
                    mBtnGetLocationUpdates.setText(R.string.start_location_updates);
                    mBtnGetLocationUpdates.setSelected(false);
                    mLocationManager.stopRequestLocation();
                    mLocationUpdates = "";
                } else {
                    mTxtInfoUpdates.setText("Getting location...");
                    v.setEnabled(false);
                    getLocationUpdates();
                }
            }
        });
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mLocationManager = new LocationManager();
        mLocationManager.onCreate(this);
    }

    @Override
    public void onDestroy() {
        mLocationManager.onDestroy();
        super.onDestroy();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        mLocationManager.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        mLocationManager.onActivityResult(requestCode, resultCode);
    }

    private void getLocation() {
        mLocationManager.requestSingleLocation(false, new OnLocationManager() {

            @Override
            public void onLocationFound(Location location, boolean isLastKnowLocation) {
                mTxtInfo.setText(getLocation(location));
                mBtnGetLocation.setEnabled(true);
            }

            @Override
            public void onLocationError(LocationError locationError) {
                switch (locationError) {
                    case DISABLED:
                        mTxtInfo.setText("Error: Location disabled");
                        break;
                    case TIMEOUT:
                        mTxtInfo.setText("Error: Timeout getting location");
                        break;
                    case REQUEST_UPDATES_ENABLED:
                        mTxtInfo.setText("Error: Request Updates are enabled");
                        break;
                    default:
                        break;
                }

                mBtnGetLocation.setEnabled(true);
            }

            @Override
            public void onPermissionsDenied() {
                mTxtInfo.setText("Permissions Denied");
                mBtnGetLocation.setEnabled(true);
            }

            @Override
            public void onProviderEnabled() {
                // Intended.
            }

            @Override
            public void onProviderDisabled() {
                // Intended.
            }
        });
    }

    private void getLocationUpdates() {
        mLocationManager.requestLocationUpdates(new OnLocationManager() {
            @Override
            public void onLocationFound(Location location, boolean isLastKnowLocation) {

                if (!mBtnGetLocationUpdates.isSelected()) {
                    mBtnGetLocationUpdates.setText(R.string.stop_location_updates);
                    mBtnGetLocationUpdates.setSelected(true);
                }

                mLocationUpdates += getLocation(location) + "\n";

                mTxtInfoUpdates.setText(mLocationUpdates);
                mBtnGetLocationUpdates.setEnabled(true);
            }

            @Override
            public void onLocationError(LocationError locationError) {
                switch (locationError) {
                    case DISABLED:
                        mTxtInfoUpdates.setText("Error: Location disabled");
                        break;
                    case TIMEOUT:
                        mTxtInfoUpdates.setText("Error: Timeout getting location");
                        break;
                    default:
                        break;
                }

                mBtnGetLocationUpdates.setEnabled(true);
            }

            @Override
            public void onPermissionsDenied() {
                mTxtInfoUpdates.setText("Permissions Denied");
                mBtnGetLocationUpdates.setEnabled(true);
            }

            @Override
            public void onProviderEnabled() {
                // Intended.
            }

            @Override
            public void onProviderDisabled() {
                // Intended.
            }
        });
    }

    private String getLocation(Location location) {
        return "Location found " + location.getLatitude() + ", " + location.getLongitude();
    }
}

File: sample-kotlin/src/main/java/com/massivedisaster/adal/samplekotlin/feature/location/FragmentLocation.kt
/*
 * ADAL - A set of Android libraries to help speed up Android development.
 *
 * Copyright (c) 2018 ADAL
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.massivedisaster.adal.samplekotlin.feature.location

import android.content.Intent
import android.location.Location
import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import com.massivedisaster.adal.fragment.BaseFragment
import com.massivedisaster.adal.samplekotlin.R
import com.massivedisaster.location.LocationManager
import com.massivedisaster.location.listener.OnLocationManager
import com.massivedisaster.location.utils.LocationError

class FragmentLocation : BaseFragment() {

    private var mLocationManager: LocationManager? = null

    private var mBtnGetLocation: Button? = null
    private var mBtnGetLocationUpdates: Button? = null
    private var mTxtInfo: TextView? = null
    private var mTxtInfoUpdates: TextView? = null
    private var mLocationUpdates = ""

    override fun getFromBundle(bundle: Bundle) {
        // Intended.
    }

    override fun restoreInstanceState(savedInstanceState: Bundle?) {
        // Intended.
    }

    override fun layoutToInflate(): Int {
        return R.layout.fragment_location
    }

    override fun doOnCreated() {
        activity!!.setTitle(R.string.sample_location)

        mBtnGetLocation = findViewById(R.id.btnGetLocation)
        mTxtInfo = findViewById(R.id.txtInfo)

        mBtnGetLocationUpdates = findViewById(R.id.btnGetLocationUpdates)
        mTxtInfoUpdates = findViewById(R.id.txtInfoUpdates)

        initialize()
    }

    fun initialize() {
        mBtnGetLocation!!.setOnClickListener { v ->
            mTxtInfo!!.text = "Getting location..."
            v.isEnabled = false
            getLocation()
        }
        mBtnGetLocationUpdates!!.setOnClickListener({ v ->
            if (v.isSelected) {
                mBtnGetLocationUpdates!!.setText(R.string.start_location_updates)
                mBtnGetLocationUpdates!!.isSelected = false
                mLocationManager!!.stopRequestLocation()
                mLocationUpdates = ""
            } else {
                mTxtInfoUpdates!!.text = "Getting location..."
                v.isEnabled = false
                getLocationUpdates()
            }
        })
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mLocationManager = LocationManager()
        mLocationManager!!.onCreate(this)
    }

    override fun onDestroy() {
        mLocationManager!!.onDestroy()
        super.onDestroy()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        mLocationManager!!.onRequestPermissionsResult(requestCode, permissions, *grantResults)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)
        mLocationManager!!.onActivityResult(requestCode, resultCode)
    }

    private fun getLocation() {
        mLocationManager!!.requestSingleLocation(false, object : OnLocationManager {

            override fun onLocationFound(location: Location, isLastKnowLocation: Boolean) {
                mTxtInfo!!.text = getLocation(location)
                mBtnGetLocation!!.isEnabled = true
            }

            override fun onLocationError(locationError: LocationError) {
                when (locationError) {
                    LocationError.DISABLED -> mTxtInfo!!.text = "Error: Location disabled"
                    LocationError.TIMEOUT -> mTxtInfo!!.text = "Error: Timeout getting location"
                    LocationError.REQUEST_UPDATES_ENABLED -> mTxtInfo!!.text = "Error: Request Updates are enabled"
                    else -> {
                    }
                }

                mBtnGetLocation!!.isEnabled = true
            }

            override fun onPermissionsDenied() {
                mTxtInfo!!.text = "Permissions Denied"
                mBtnGetLocation!!.isEnabled = true
            }

            override fun onProviderEnabled() {
                // Intended.
            }

            override fun onProviderDisabled() {
                // Intended.
            }

            override fun onStopRequestUpdate() {

            }
        })
    }

    private fun getLocationUpdates() {
        mLocationManager!!.requestLocationUpdates(object : OnLocationManager {
            override fun onLocationFound(location: Location, isLastKnowLocation: Boolean) {

                if (!mBtnGetLocationUpdates!!.isSelected) {
                    mBtnGetLocationUpdates!!.setText(R.string.stop_location_updates)
                    mBtnGetLocationUpdates!!.isSelected = true
                }

                mLocationUpdates += getLocation(location) + "\n"

                mTxtInfoUpdates!!.text = mLocationUpdates
                mBtnGetLocationUpdates!!.isEnabled = true
            }

            override fun onLocationError(locationError: LocationError) {
                when (locationError) {
                    LocationError.DISABLED -> mTxtInfoUpdates!!.text = "Error: Location disabled"
                    LocationError.TIMEOUT -> mTxtInfoUpdates!!.text = "Error: Timeout getting location"
                    else -> {
                    }
                }

                mBtnGetLocationUpdates!!.setEnabled(true)
            }

            override fun onPermissionsDenied() {
                mTxtInfoUpdates!!.text = "Permissions Denied"
                mBtnGetLocationUpdates!!.isEnabled = true
            }

            override fun onProviderEnabled() {
                // Intended.
            }

            override fun onProviderDisabled() {
                // Intended.
            }

            override fun onStopRequestUpdate() {
                mLocationUpdates += "STOP - Request Updates"
                mTxtInfoUpdates!!.setText(mLocationUpdates)
            }
        })
    }

    private fun getLocation(location: Location): String {
        return "Location found " + location.latitude + ", " + location.longitude
    }
}